### PR TITLE
add(scan): Implement gRPC support for registering viewing keys for scanning

### LINKED ISSUE #8176: add(scan): `RegisterKeys` gRPC
### Motivation

We want to be able to register viewing keys over the gRPC interface.

Depends on #8203.

Tracked in #7728.

### PR DESCRIPTION
**Pull request by upbqdn**
## Motivation

Close #8176 

#### Check before marking the PR as ready for review:
  - [x] Will the PR name make sense to users?
  - [x] Does the PR have a priority label?
  - [ ] Have you added or updated tests?
  - [x] Is the documentation up to date?

## Solution

- [Use use tonic::Request for consistency](https://github.com/ZcashFoundation/zebra/commit/1437712ed10882a710002428e1cbfe5b6c468f02)
- [Add definitions of proto types](https://github.com/ZcashFoundation/zebra/commit/64a3433307628addce5f759e7e7944baf836c1a3)
- [Impl register_keys gRPC](https://github.com/ZcashFoundation/zebra/commit/39fb2b1addb1929d3d23882c60f4947de4f5feb2)

### Testing

We'll start with tests in #8244.

### Reviewer Checklist

Check before approving the PR:
  - [ ] Does the PR scope match the ticket?
  - [ ] Are there enough tests to make sure it works? Do the tests cover the PR motivation?
  - [ ] Are all the PR blockers dealt with?
        PR blockers can be dealt with in new tickets or PRs.

_And check the PR Author checklist is complete._

## Follow Up Work

- #8244
- #8272

### CODE REVIEW

**Comment by oxarbitrage** [comment]
It seems we have a problem with the new optional field in the proto file https://github.com/ZcashFoundation/zebra/actions/runs/7888805960/job/21527160060?pr=8266#step:10:1572

**Comment by upbqdn** [comment]
We shouldn't see that error since `Protocol Buffers v3.15.0`, which was released back in 2021. Look at the first entry in the changelog: https://github.com/protocolbuffers/protobuf/releases/tag/v3.15.0. The issue describing the error was closed a long time ago: https://github.com/google/protobuf-gradle-plugin/issues/415.

@gustavovalverde could you please have a look? It looks like we're using version pre 3.15.0?

**Comment by upbqdn** [comment]
I couldn't reproduce the error locally.

**Comment by upbqdn** [comment]
The error is: 

```
protoc failed: proto/scanner.proto: This file contains proto3 optional fields, but --experimental_allow_proto3_optional was not set.
```

**Comment by arya2** [comment]
Try using `oneof` with only one option, it worked for me earlier. ~(It's an issue with prost, not protoc)~

Update: I was using an outdated version of protoc locally.

File: zebra-grpc/proto/scanner.proto [hunk 1]
```diff
@@ -62,3 +77,11 @@ message TransactionHash {
     // A transaction id hash
     repeated string hash = 1;
 }
+
+// A scanning key with an optional birth height
+message KeyWithHeight {
+    // Scanning key
+    string key = 1;
+    // Birth height of the key
+    optional uint32 height = 2;
```
Comment (arya2) [inline zebra-grpc/proto/scanner.proto:86]: We could also remove this field for now.

Comment (upbqdn) [reply zebra-grpc/proto/scanner.proto:86]: Let's see how #8273 goes.

**Comment by upbqdn** [comment]
PR #8273 should resolve the issue.

**Comment by upbqdn** [comment]
> PR https://github.com/ZcashFoundation/zebra/pull/8273 should resolve the issue.

It did.

**Comment by arya2** [comment]
Could you show the manual test output?

**Comment by arya2** [comment]
Manual test output:

```console

~/github/zebra$ grpcurl -plaintext -import-path ./zebra-grpc/proto -proto zebra-grpc/proto/scanner.proto -d '{"keys": [{ "key": "zxviews1q0duytgcqqqqpqre26wkl45gvwwwd706xw608hucmvfalr759ejwf7qshjf5r9aa7323zulvz6plhttp5mltqcgs9t039cx2d09mgq05ts63n8u35hyv6h9nc9ctqqtue2u7cer2mqegunuulq2luhq3ywjcz35yyljewa4mgkgjzyfwh6fr6jd0dzd44ghk0nxdv2hnv4j5nxfwv24rwdmgllhe0p8568sgqt9ckt02v2kxf5ahtql6s0ltjpkckw8gtymxtxuu9gcr0swvz" }]}' '127.0.0.1:8231' scanner.Scanner/RegisterKeys
{
  "keys": [
    "zxviews1q0duytgcqqqqpqre26wkl45gvwwwd706xw608hucmvfalr759ejwf7qshjf5r9aa7323zulvz6plhttp5mltqcgs9t039cx2d09mgq05ts63n8u35hyv6h9nc9ctqqtue2u7cer2mqegunuulq2luhq3ywjcz35yyljewa4mgkgjzyfwh6fr6jd0dzd44ghk0nxdv2hnv4j5nxfwv24rwdmgllhe0p8568sgqt9ckt02v2kxf5ahtql6s0ltjpkckw8gtymxtxuu9gcr0swvz"
  ]
}

```

Output for an already-registered key:

```console

~/github/zebra$ grpcurl -plaintext -import-path ./zebra-grpc/proto -proto zebra-grpc/proto/scanner.proto -d '{"keys": ["zxviews1q0duytgcqqqqpqre26wkl45gvwwwd706xw608hucmvfalr759ejwf7qshjf5r9aa7323zulvz6plhttp5mltqcgs9t039cx2d09mgq05ts63n8u35hyv6h9nc9ctqqtue2u7cer2mqegunuulq2luhq3ywjcz35
yyljewa4mgkgjzyfwh6fr6jd0dzd44ghk0nxdv2hnv4j5nxfwv24rwdmgllhe0p8568sgqt9ckt02v2kxf5ahtql6s0ltjpkckw8gtymxtxuu9gcr0swvz"]}' '127.0.0.1:8231' scanner.Scanner/DeleteKeys
{}


```

**Review by arya2 — APPROVED** [review]
Looks good!

**Pull request merged**